### ISSUE TITLE
How to create zip in memory or by providing FILE* stream?

### ISSUE THREAD

**Issue by bruenor41** [nih-at/libzip#372]
Hi,

Systems are more and more restricted and I work with restricted android public storage. Libzip can't work directly with the file system using paths. However, host application can tell system to create file and get FILE* stream for blank binary file. 

Is possible provide FILE* as target for newly created zip, instead path? FILE* should be truncated at beginning of process.

If no, second option is to create zip in memory and then save buffer to FILE*. But previous option would be better.

I look at this example
https://github.com/nih-at/libzip/blob/main/examples/in-memory.c

In example is created "zip_source_t *src;" using "zip_source_buffer_create(data, size, 1, &error)". It is filled with values from "getData", because first operation in example is reading zip from file. Then is "zip_source_t" preserved to second step "zip in memory".

But I am confused, how to create "zip_source_t" only for zipping?

**Comment by dillof** [comment]
libzip does not support writing to `FILE*` out of the box, but you could create a source via `zip_source_function` that implements it. 

For example, you could 
- return 0 for `ZIP_SOURCE_READ` (empty file)
- do nothing but return 0 (no error) on `ZIP_SOURCE_OPEN`, `ZIP_SOURCE_CLOSE`, and `ZIP_SOURCE_BEGIN_WRITE`
- truncate the file on `ZIP_SOURCE_ROLLBACK`
- pass `ZIP_SOURCE_WRITE`, `ZIP_SOURCE_SEEK_WRITE`, and `ZIP_SOURCE_TELL_WRITE` on to `fwrite`, `fseek`, and `ftell` respectively
- close the file on `ZIP_SOURCE_FREE`

**Comment by bruenor41** [comment]
Thank you very much fast response. I will try it to put together like you wrote. Shame there is no example. Btw, I found that unzip process is already implemented for restrictes storage, there exists "zip_fdopen".

**Comment by dillof** [comment]
Since we can't atomically replace a `FILE*`, `zip_fdopen` does not support writing. We could add the concept of write-only archives, which we could support for `FILE*` if it's empty or `ZIP_TRUNCATE` is set.

If you write a short program implementing such a source, we'll gladly include it in the distribution as an example.

**Comment by QrczakMK** [comment]
I am considering the following semantics in a similar case I have (not with `FILE*`, but I will state it in the language of `FILE*`, and for simplicity assume that the initial position was 0):

If the `FILE*` is write-only, pretend that the archive does not exist: let `ZIP_SOURCE_STAT` return `ZIP_ER_READ` / `ENOENT`. In this case libzip should not call `ZIP_SOURCE_OPEN`, which would fail, but for robustness `ZIP_ER_OPEN` / `ENOENT` would be returned in that case.

If the `FILE* supports both reading and writing, let `ZIP_SOURCE_BEGIN_WRITE` and `ZIP_SOURCE_BEGIN_WRITE_CLONING` decide between two strategies:
1. If the `FILE*` is empty, let libzip write directly to it. `ZIP_SOURCE_COMMIT_WRITE` will do nothing. `ZIP_SOURCE_ROLLBACK_WRITE` will truncate the file to empty.
2. Otherwise, let libzip accumulate data in memory. `ZIP_SOURCE_COMMIT_WRITE` will truncate the file to empty, write data from memory, and clear memory. `ZIP_SOURCE_ROLLBACK_WRITE` clear memory.

In either case `ZIP_SOURCE_REMOVE` truncates the file to empty.

Truncation is not really supported for `FILE*`. It might be implemented using `fileno()` and `ftruncate()` on Unix / `_chsize_s()` on Windows if they are supported. In any case truncation is not attempted if the write position (maintained separately from the read position) is 0 and truncation is not actually supported.

Please tell if you find flaws in that.

**Comment by bruenor41** [comment]
I am noy author, but it looks like I need. Can you please share then example?

**Comment by dillof** [comment]
> I am considering the following semantics in a similar case I have (not with `FILE*`, but I will state it in the language of `FILE*`, and for simplicity assume that the initial position was 0):
> 
> If the `FILE*` is write-only, pretend that the archive does not exist: let `ZIP_SOURCE_STAT` return `ZIP_ER_READ` / `ENOENT`. In this case libzip should not call `ZIP_SOURCE_OPEN`, which would fail, but for robustness `ZIP_ER_OPEN` / `ENOENT` would be returned in that case.
> 
> If the `FILE*` supports both reading and writing, let `ZIP_SOURCE_BEGIN_WRITE` and `ZIP_SOURCE_BEGIN_WRITE_CLONING` decide between two strategies:
> 
> 1. If the `FILE*` is empty, let libzip write directly to it. `ZIP_SOURCE_COMMIT_WRITE` will do nothing. `ZIP_SOURCE_ROLLBACK_WRITE` will truncate the file to empty.

If the application is killed while writing the file, an incomplete copy of the new archive is left with the final file name; it might not be obvious for the user that the written file is incomplete.

> 2. Otherwise, let libzip accumulate data in memory. `ZIP_SOURCE_COMMIT_WRITE` will truncate the file to empty, write data from memory, and clear memory. `ZIP_SOURCE_ROLLBACK_WRITE` clear memory.

If the application is killed while writing the file, the existing archive is replaced with an incomplete copy of the new archive, which results in data loss. 

Both of these cases violate one of the design criteria of libzip, which is why libzip does not support them. 

It might be acceptable for your application, in which case by all means implement it. Just be aware and clearly document that data loss is a possibility.

**Comment by bruenor41** [comment]
For example android offers cache path where libzip is capable to work with standard paths. Maybe it can be used as temp path. Final destination is restricted public storage. In my case I can provide temp location in memory or in cache, but destination file is on public storage and I must get FILE* from restricted system, like I described in my first post. Just now I fully create zip in cache and then I copy it to FILE*. But this is double work and it is used in case when I really need to do it quickly.

**Comment by QrczakMK** [comment]
Yes, I am aware that this does not provide atomicity. Also, keeping the temporary destination separate from the final destination is needed for updating an existing archive, as we might copy data from the original archive, from positions corresponding to regions already written in the temporary archive.

This is why for named local files I will recommend a different zip source, which works like builtin libzip handling of files: creating a temporary file and renaming, only using file descriptors with custom buffering logic instead of stdio (since this is [already implemented for other reasons](https://github.com/google/riegeli/blob/master/riegeli/bytes/fd_writer.h)).

There will be a similar source for our internal (not open sourced) `File` abstraction, which generalizes local files and often supports renaming.

Still this is not general enough — not all destinations support the notion of temporary files with renaming. And sometimes atomicity does not matter at all, e.g. when composing the archive in memory. So I will also provide a zip source which writes to an arbitrary `riegeli::Writer` in the way I described above (and [`riegeli::CFileWriter`](https://github.com/google/riegeli/blob/master/riegeli/bytes/cfile_writer.h) wraps `FILE*` as `riegeli::Writer`). I have pending usages which compose an archive in memory.

* * *

BTW, the name “zip source” is not great when we deal with writing. The name “source” suggests reading only.

So far I kept the libzip terminology in this respect. The wrapping class is tentatively named `riegeli::ZipSource`, with concrete values created by `riegeli::FdZipSource(filename)`, `riegeli::FileZipSource(filename)`, `riegeli::ReaderZipSource(riegeli::Reader)`, `riegeli::WriterZipSource(riegeli::Writer)`, or by deriving from `riegeli::ZipSourceAdapter` and overriding particular operations.

`riegeli::ZipSource` is used for an archive (reading and writing), for reading existing entries (reading), and for providing new contents of entries (reading too, since libzip is reading from what the client provides). There are other mechanisms for dealing with entries for convenience, in particular writing by the client to a `riegeli::Writer` which buffers data in memory and updates the entry on `Close()`.

Perhaps some other name instead of `riegeli::ZipSource` would be better.
